syntax = "proto3";

message _display {
    int32 layout                = 1;
    int32 alt_layout            = 2;

    _small small                = 3;
    message _small {
        int32 alpha             = 1;
        int32 border            = 2;
        int32 position          = 3;
    }
}

message _cpu {
    _freq freq                  = 1;
    message _freq {
        int32 min               = 1;
        int32 max               = 2;
    }

    _core core                  = 2;
    message _core {
        int32 min               = 1;
        int32 max               = 2;
    }
}

message _pen {
    string image                = 1;
    bool screen0                = 2;

    _speed speed                = 3;
    message _speed {
        int32 x                 = 1;
        int32 y                 = 2;
    }
}

message _menu {
    string bg                   = 1;
    bool show_cursor            = 2;
}

message _autosave {
    bool enable                 = 1;
    int32 slot                  = 2;
}

message _keypad {
    int32 rotate                = 1;

    _hotkey hotkey              = 2;
    enum _hotkey {
        menu                    = 0;
        select                  = 1;
    }

    _swap swap                  = 3;
    message _swap {
        bool l1_l2              = 1;
        bool r1_r2              = 2;
    }
}

message _joystick {
    _mode mode                  = 1;
    enum _mode {
        keypad                  = 0;
        pen                     = 1;
        remap                   = 2;
    }

    int32 dead_zone             = 2;

    _remap remap_left           = 3;
    _remap remap_right          = 4;
    message _remap {
        int32 top               = 1;
        int32 down              = 2;
        int32 left              = 3;
        int32 right             = 4;
    }
}

message settings {
    string version              = 1;
    string language             = 2;
    string font_path            = 3;
    string state_folder         = 4;
    string border_image         = 5;

    int32 debug_level           = 6;
    int32 system_volume         = 7;
    int32 fast_forward          = 8;

    bool half_volume            = 9;
    bool low_battery_close      = 10;

    _display display            = 11;
    _cpu cpu                    = 12;
    _pen pen                    = 13;
    _menu menu                  = 14;
    _autosave autosave          = 15;
    _keypad keypad              = 16;
    _joystick joystick          = 17;
}

